packages/tpetra/core/src/Tpetra_CrsGraph_def.hpp

```diff
@@ -6084,20 +6092,41 @@ namespace Tpetra {
       "Import or Export operations are not allowed on the destination "
       "CrsGraph if it is fill complete.");
 
-    const map_type& rowMap = * (this->rowMap_);
     const size_t numImportLIDs = static_cast<size_t> (importLIDs.size ());
+    
+    // If we're inserting in local indices, let's pre-allocate
+    Teuchos::Array<LO> lclColInds;
+    if(this->isLocallyIndexed()) { 
+      size_t maxNumInserts = 0;
+      for (size_t i = 0; i < numImportLIDs; ++i) 
+        maxNumInserts = std::max(maxNumInserts,numPacketsPerLID[i]);
+      lclColInds.resize(maxNumInserts);
+    }
+
+    const map_type& rowMap = * (this->rowMap_);
     size_t importsOffset = 0;
     for (size_t i = 0; i < numImportLIDs; ++i) {
       const LO lclRow = importLIDs[i];
       const GO gblRow = rowMap.getGlobalElement (lclRow);
       const LO numEnt = numPacketsPerLID[i];
       const GO* const gblColInds = (numEnt == 0) ? nullptr : &imports[importsOffset];
-      if (gblRow == Tpetra::Details::OrdinalTraits<GO>::invalid ()) {
-        // This row is not in the row Map on the calling process.
-        this->insertGlobalIndicesIntoNonownedRows (gblRow, gblColInds, numEnt);
-      }
+      if(!this->isLocallyIndexed()) {
+        if (gblRow == Tpetra::Details::OrdinalTraits<GO>::invalid ()) {
+          // This row is not in the row Map on the calling process.
+          this->insertGlobalIndicesIntoNonownedRows (gblRow, gblColInds, numEnt);
```

Comment:
Do we ever reach this case where the incoming row index is not in the row Map on the calling (in this case, receiving) process?  That doesn't really make sense -- if the Export or Import directed those rows to go to this process, then those rows should be in the target CrsGraph's row Map, no?